### PR TITLE
Integrate with Service Connect bridge-mode CNI plugin

### PR DESCRIPTION
**Pull request by yinyic**
### Summary

ServiceConnect-enabled tasks require network namespace configurations that are performed by a new CNI plugin. This PR integrates with the new CNI plugin `ecs-serviceconnect` for bridge-mode tasks.

### Implementation details

In the previous PR https://github.com/aws/amazon-ecs-agent/pull/3255 where we integrated with SC CNI plugin for awsvpc tasks we've already added new SC CNI config models. 

In this PR, a new `RedirectMode` CNI config parameter is introduced. The value will be `NAT` and `TPROXY` for awsvpc and bridge-mode tasks respectively. Additionally, a `RedirectIP` field is introduced for further distinguishing one of the two scenarios for bridge mode netns setup.

For bridge-mode SC-enabled tasks, one pause container will be created for each task container, including both appnet and customer container. Depending on CNI configuration parameters the plugin will do one of two things
1. If `EgressConfig.RedirectIP` is included, create an IP route to redirect traffic destined to SC VIPs to the specified redirect IP (which will be AppNet container IP). **This is the configuration for customer pause containers.**
2. If `EgressConfig.ListenerPort` is included, set up a TPROXY. **This is the configuration for AppNet pause containers.**

In order for other task pause containers to be able to access AppNet container IP (i.e. the `RedirectIP`) during their netns setup. A `PauseContainerIPConfig` property is added to service connect `NetworkConfig`. The IP addresses will be populated once AppNet pause container is started.

TODO - Update engine `TestContainersWithServiceConnect_BridgeMode` test to be more reliable (https://github.com/aws/amazon-ecs-agent/pull/3262)

### Testing

New tests cover the changes:  yes

### Description for the changelog

Integrate with Service Connect bridge-mode CNI plugin

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/api/task/task.go [hunk 1]
```diff
@@ -1539,9 +1539,9 @@ func (task *Task) GetBridgeModePauseContainerForTaskContainer(container *apicont
 	return pauseContainer, nil
 }
 
-// getBridgeModeTaskContainerForPauseContainer retrieves the associated task container for a pause container in a bridge-mode SC-enabled task.
+// GetBridgeModeTaskContainerForPauseContainer retrieves the associated task container for a pause container in a bridge-mode SC-enabled task.
 // For a container with name "abc", the pause container will always be named "~internal~ecs~pause-abc"
-func (task *Task) getBridgeModeTaskContainerForPauseContainer(container *apicontainer.Container) (*apicontainer.Container, error) {
+func (task *Task) GetBridgeModeTaskContainerForPauseContainer(container *apicontainer.Container) (*apicontainer.Container, error) {
```
Comment (aws-gibbskt) [inline agent/api/task/task.go:1544]: Where is this accessed out of package? I don't know why this change is needed.

File: agent/api/serviceconnect/service_connect.go [hunk 2]
```diff
@@ -32,6 +32,8 @@ type RuntimeConfig struct {
 	StatsRequest string `json:"statsRequest"`
 	// HTTP Path + Params to drain ServiceConnect connections
 	DrainRequest string `json:"drainRequest"`
+	// SC pause container IP address - used for bridge-mode CNI configuration
+	PauseContainerIPConfig *PauseContainerIPConfig `json:"pauseContainerIPConfig,omitempty"`
```
Comment (rajal-amzn) [inline agent/api/serviceconnect/service_connect.go:36]: Would you prefer storing these as `NetworkConfig` along with pause container associated with customer containers ?

File: agent/api/task/task_linux.go [hunk 3]
```diff
@@ -355,3 +357,38 @@ func (task *Task) BuildCNIConfig(includeIPAMConfig bool, cniConfig *ecscni.Confi
 
 	return cniConfig, nil
 }
+
+// BuildCNIConfigBridgeMode builds a list of CNI network configurations for a task in docker bridge mode.
+func (task *Task) BuildCNIConfigBridgeMode(cniConfig *ecscni.Config, containerName string) (*ecscni.Config, error) {
```
Comment (rajal-amzn) [inline agent/api/task/task_linux.go:362]: nit: I see all the methods are marked public. why?

File: agent/api/task/task_linux.go [hunk 4]
```diff
@@ -355,3 +357,38 @@ func (task *Task) BuildCNIConfig(includeIPAMConfig bool, cniConfig *ecscni.Confi
...
+	if !task.IsNetworkModeBridge() {
+		return nil, errors.New("non bridge-mode task should not invoke BuildCNIConfigBridgeMode")
+	}
+
+	var netconf *libcni.NetworkConfig
+	var ifName string
+	var err error
+
+	// Build a CNI network configuration for bridge-mode SC-enabled task
+	if task.IsServiceConnectEnabled() {
```
Comment (rajal-amzn) [inline agent/api/task/task_linux.go:372]: If evaluated to false, shouldnt we be retuning error?

File: agent/ecscni/netconfig_linux.go [hunk 5]
```diff
@@ -176,6 +176,8 @@ func NewAppMeshConfig(appMesh *appmesh.AppMesh, cfg *Config) (string, *libcni.Ne
 // NewServiceConnectNetworkConfig creates a new ServiceConnect CNI network configuration
 func NewServiceConnectNetworkConfig(
 	scConfig *serviceconnect.Config,
+	redirectMode RedirectMode,
+	shouldIncludeRedirectIP bool,
```
Comment (chienhanlin) [inline agent/ecscni/netconfig_linux.go:180]: nit. `shouldIncludeRedirectIP, enableIPv4, enableIPv6 bool,`  there of them are all boolean.

File: agent/api/task/task_linux_test.go [hunk 6]
```diff
@@ -67,6 +70,16 @@ const (
...
+
+	scContainerName      = "service-connect"
+	scEgressListenerPort = 12345
+	scInterceptPort      = 8080
+	scListenerPort       = 15000
+	scPauseIPv4          = "172.0.0.2"
+)
+
+var (
+	scPauseContainerName = fmt.Sprintf("%s-%s", NetworkPauseContainerName, scContainerName)
```
Comment (chienhanlin) [inline agent/api/task/task_linux_test.go:82]: nit. feel that we use this in more than one place. Should we turn this into a const template in task.go and reuse it in following lines?
* https://github.com/aws/amazon-ecs-agent/blob/feature/service-connect/agent/api/task/task.go#L1504
* https://github.com/aws/amazon-ecs-agent/blob/feature/service-connect/agent/api/task/task.go#L1534

File: agent/api/task/task_linux.go (hunk 3 above)
Comment (chienhanlin) [reply agent/api/task/task_linux.go:362]: Making `BuildCNIConfigAwsvpc` and `BuildCNIConfigBridgeMode` public and than they can be used in the engine package, like:

`agent/engine/docker_task_engine.go`
* buildCNIConfigFromTaskContainerAwsvpc() [[L1940](https://github.com/yinyic/amazon-ecs-agent/blob/6a66b11cbc4f727012a78b35470cad773e836454/agent/engine/docker_task_engine.go#L1940)]

* buildCNIConfigFromTaskContainerBridgeMode() [[L1958](https://github.com/yinyic/amazon-ecs-agent/blob/6a66b11cbc4f727012a78b35470cad773e836454/agent/engine/docker_task_engine.go#L1958)]

`agent/engine/task_manager.go` 
* releaseIPInIPAM()[[L757](https://github.com/yinyic/amazon-ecs-agent/blob/6a66b11cbc4f727012a78b35470cad773e836454/agent/engine/task_manager.go#L757)]

File: agent/api/task/task.go (hunk 1 above)
Comment (yinyic) [reply agent/api/task/task.go:1544]: Thanks for calling out. This was an intermediate change when I was working on the PR. It's no longer needed. Will remove.

File: agent/api/serviceconnect/service_connect.go (hunk 2 above)
Comment (yinyic) [reply agent/api/serviceconnect/service_connect.go:36]: Do you mean storing it directly inside `ecscni.ServiceConnectConfig` [struct](https://github.com/aws/amazon-ecs-agent/blob/feature/service-connect/agent/ecscni/types_linux.go#L170)? We collect AppNet pause IPs from docker daemon, but the method that we create this config object currently does not have a handle of docker client. I'd rather keep it that way, because this config object is intended to be an API object for CNI plugin, so it probably shouldn't need to know or care about the container runtime.

File: agent/api/task/task_linux.go (hunk 3 above)
Comment (yinyic) [reply agent/api/task/task_linux.go:362]: Thanks ChienHan! 
Like ChienHan mentioned, docker engine drives the container lifecycle as a state machine, and task package owns most of the actions to take during each task/container state.

File: agent/api/task/task_linux.go (hunk 4 above)
Comment (yinyic) [reply agent/api/task/task_linux.go:372]: I was thinking that we may add other bridge-mode CNI plugins in the future so I kept it open. But you are right that we probably should throw error because SC task is the only one that should be invoking this method at this moment.

File: agent/ecscni/netconfig_linux.go (hunk 5 above)
Comment (yinyic) [reply agent/ecscni/netconfig_linux.go:180]: Will update to ist each in its own line with an explicit type.

File: agent/api/task/task.go [hunk 7]
```diff
@@ -3203,6 +3203,15 @@ func (task *Task) GetServiceConnectContainer() *apicontainer.Container {
 	return c
 }
 
+func (task *Task) IsContainerServiceConnectPause(containerName string) bool {
```
[the remark this answers is not in the document]
Comment (yinyic) [reply agent/api/task/task.go:3206]: Will add.

File: agent/api/task/task_linux_test.go (hunk 6 above)
Comment (yinyic) [reply agent/api/task/task_linux_test.go:82]: Will do

File: agent/api/task/task_linux_test.go [hunk 8]
```diff
@@ -1316,6 +1329,54 @@ func TestBuildCNIConfigRegularENIWithAppMesh(t *testing.T) {
...
+	for _, blockIMDS := range []bool{true, false} {
+		t.Run(fmt.Sprintf("When BlockInstanceMetadata is %t", blockIMDS), func(t *testing.T) {
+			testTask := &Task{}
+			testTask.AddTaskENI(getTestENI())
+			testTask.ServiceConnectConfig = &serviceconnect.Config{
+				ContainerName: scContainerName,
+				IngressConfig: []serviceconnect.IngressConfigEntry{{ListenerPort: scListenerPort}},
+				EgressConfig:  &serviceconnect.EgressConfig{ListenerPort: scEgressListenerPort},
+			}
+			testTask.Containers = []*apicontainer.Container{{Name: "service-connect"}}
```
[the remark this answers is not in the document]
Comment (yinyic) [reply agent/api/task/task_linux_test.go:1342]: Will do

File: agent/api/task/task.go [hunk 9]
```diff
@@ -3203,6 +3205,18 @@ func (task *Task) GetServiceConnectContainer() *apicontainer.Container {
...
 
+// IsContainerServiceConnectPause checks whether a given container name is the name of the task service connect pause
+// container. We construct the name of SC pause container by taking SC container name from SC config, and using the
+// pause container naming pattern.
+func (task *Task) IsContainerServiceConnectPause(containerName string) bool {
+	scContainer := task.GetServiceConnectContainer()
+	if scContainer == nil {
+		return false
+	}
+	scPauseName := fmt.Sprintf("%s-%s", NetworkPauseContainerName, scContainer.Name)
```
Comment (chienhanlin) [inline agent/api/task/task.go:3216]: nit and sorry I missed this line in my previous comment.
```
scPauseName := fmt.Sprintf(ServiceConnectPauseContainerNameFormat, scContainer.Name)
```

File: agent/api/task/task_linux.go [hunk 10]
```diff
@@ -355,3 +357,37 @@ func (task *Task) BuildCNIConfig(includeIPAMConfig bool, cniConfig *ecscni.Confi
...
+func (task *Task) BuildCNIConfigBridgeMode(cniConfig *ecscni.Config, containerName string) (*ecscni.Config, error) {
+	if !task.IsNetworkModeBridge() || !task.IsServiceConnectEnabled() {
+		return nil, errors.New("only bridge-mode Service-Connect-enabled task should invoke BuildCNIConfigBridgeMode")
+	}
+
+	var netconf *libcni.NetworkConfig
+	var ifName string
+	var err error
+
+	scPauseIPConfig := task.GetServiceConnectRuntimeConfig().PauseContainerIPConfig
```
Comment (chienhanlin) [inline agent/api/task/task_linux.go:372]: Same Q asked in https://github.com/aws/amazon-ecs-agent/pull/3265/files#r914210160. Thanks!

File: agent/engine/docker_task_engine.go [hunk 11]
```diff
@@ -1752,26 +1781,72 @@ func (engine *DockerTaskEngine) provisionContainerResources(task *apitask.Task,
...
+
+	log.Infof("Successfully configured pause netns for %s", container.Name)
+	return dockerapi.MetadataFromContainer(containerInspectOutput)
+}
+
+// checkTearDownPauseContainer idempotently tears down the pause container network when the pause container's known
 //or desired status is stopped.
-func (engine *DockerTaskEngine) checkTearDownPauseContainerAwsvpc(task *apitask.Task) {
-	if !task.IsNetworkModeAWSVPC() {
+func (engine *DockerTaskEngine) checkTearDownPauseContainer(task *apitask.Task) {
```
Comment (fierlion) [inline agent/engine/docker_task_engine.go:1840]: nit: to tie to the pause container naming thread -- here you're not differentiating between the sc pause container and the base pause container.  If this should only apply to the ServiceConnect specific pause container I would call this `checkTearDownServiceConnectPauseContainer()`... 

however my preference (as I noted above) is just reusing the `NetworkPauseContainerName` and keeping it generic.

File: agent/api/task/task.go (hunk 9 above)
Comment (yinyic) [reply agent/api/task/task.go:3216]: Thanks! Will update

File: agent/ecscni/netconfig_linux.go [hunk 12]
```diff
@@ -192,12 +195,37 @@ func NewServiceConnectNetworkConfig(
...
+		case NAT:
+			// NAT redirect mode is for awsvpc tasks, where the one and only pause container netns will have a NAT redirect rule.
+			egressConfig.ListenerPort = scConfig.EgressConfig.ListenerPort
+		case TPROXY: // bridge
+			// TPROXY redirect mode is used for bridge-mode tasks. There are two use cases:
+			// 1. SC pause container netns will set up TPROXY that requires the Egress port
+			// 2. Other task pause container netns will add a route for traffic destined for SC VIP-CIDR to go to SC container.
+			//    In that case the configuration requires the SC (pause) container IP.
+			if shouldIncludeRedirectIP {
+				scPauseIPConfig := scConfig.RuntimeConfig.PauseContainerIPConfig
```
[the remark this answers is not in the document]
Comment (yinyic) [reply agent/ecscni/netconfig_linux.go:214]: That's a good callout - no it won't be null if we already get here, but it might be null when we populate pause container IPs so I'll add a check there.

File: agent/api/task/task_linux.go (hunk 10 above)
Comment (yinyic) [reply agent/api/task/task_linux.go:372]: it won't be null if we already get here, but it might be null when we populate pause container IPs in `task.PopulateServiceConnectPauseIPConfig` - I'll make sure to create `RuntimeConfig` if it does not already exist

File: agent/api/task/task_linux_test.go [hunk 13]
```diff
@@ -1367,3 +1428,71 @@ func TestBuildCNIConfigTrunkBranchENI(t *testing.T) {
...
+			// The first one should be for the ENI.
+			var scConfig ecscni.ServiceConnectConfig
+			err = json.Unmarshal(cniConfig.NetworkConfigs[0].CNINetworkConfig.Bytes, &scConfig)
+			require.NoError(t, err)
+			assert.Equal(t, 1, len(scConfig.IngressConfig))
+			assert.Equal(t, uint16(scListenerPort), scConfig.IngressConfig[0].ListenerPort)
+			assert.NotNil(t, scConfig.EgressConfig)
+			assert.Equal(t, string(ecscni.TPROXY), scConfig.EgressConfig.RedirectMode)
+			if containerName != scPauseContainerName {
+				// Should only include redirect IPs if container is an application pause container
```
[the remark this answers is not in the document]
Comment (yinyic) [reply agent/api/task/task_linux_test.go:1464]: > is it possible that the network pause container could ever be named the same as the service connect application pause container? 

Sorry not quite sure if I follow. What do we mean by "network pause container"?

When ECS Agent receives a bridge-mode service-connect task, and let's say there are N customer application container, and 1 service connect container. Those (N+1) containers all have unique names (validated by control plane). When we create pause containers for each of them and name the pause containers `~internal~ecs~pause-<$ACTUAL_CONTAINER_NAME>`, then it's guaranteed that all (N+1) pause containers will also have unique names.

File: agent/api/task/task_linux.go [hunk 14]
```diff
@@ -355,3 +357,37 @@ func (task *Task) BuildCNIConfig(includeIPAMConfig bool, cniConfig *ecscni.Confi
 
 	return cniConfig, nil
 }
+
+// BuildCNIConfigBridgeMode builds a list of CNI network configurations for a task in docker bridge mode.
+// Currently the only plugin in available is for Service Connect
+func (task *Task) BuildCNIConfigBridgeMode(cniConfig *ecscni.Config, containerName string) (*ecscni.Config, error) {
```
[the remark this answers is not in the document]
Comment (yinyic) [reply agent/api/task/task_linux.go:363]: I separated bridge-mode CNI config creation from the AWSVPC because they are very different - bridge-mode only needs the ServiceConnect CNI plugin, while AWSVPC need a bunch more, and other configurations. 

If we are introducing more network modes that will use CNI plugins, then we could combine it with the existing AWSVPC or bridge CNI config build if there's a lot of similarity, or we could create a new one for its own.

File: agent/ecscni/plugin_linux_test.go [hunk 15]
```diff
@@ -651,88 +657,198 @@ func TestConstructIPAMNetworkConfig(t *testing.T) {
...
-	assert.Equal(t, true, scNetworkConfig.EnableIPv4)
-	assert.Equal(t, false, scNetworkConfig.EnableIPv6)
+	testCases := []struct {
+		redirectMode            RedirectMode
+		shouldIncludeRedirectIP bool
+	}{
+		{
+			redirectMode:            NAT,
+			shouldIncludeRedirectIP: false,
+		},
```
[the remark this answers is not in the document]
Comment (yinyic) [reply agent/ecscni/plugin_linux_test.go:667]: That's not a valid use case. When building SC CNI config for AWSVPC task, we will always set redirectMode to `NAT` and `shouldIncludeRedirectIP` to false.

File: agent/engine/docker_task_engine.go (hunk 11 above)
Comment (yinyic) [reply agent/engine/docker_task_engine.go:1840]: > you're not differentiating between the sc pause container and the base pause container

That's actually the intention. The branching for networking mode comes in at a later point inside `cleanupPauseContainerNetwork` method, similar to how we branch in `provisionContainerResources`

File: agent/ecscni/netconfig_linux.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (yinyic) [reply agent/ecscni/netconfig_linux.go]: Created a new field `NetworkConfig` that's no longer nested in `RuntimeConfig` to avoid the complication that multiple containers could potentially be creating it.

File: agent/api/task/task_linux.go (hunk 10 above)
Comment (yinyic) [reply agent/api/task/task_linux.go]: Created a new field `NetworkConfig` that's no longer nested in `RuntimeConfig` to avoid the complication that multiple containers could potentially be creating it.

**Review by fierlion — DISMISSED** [review]

**Review by chienhanlin — APPROVED** [review]

**Review by rajal-amzn — COMMENTED** [review]
Changes look good. Have we also picked up the latest CNI changes? Or would you be taking that up in a separate PR.

**Review by chienhanlin — DISMISSED** [review]

**Review by fierlion — APPROVED** [review]

**Review by rajal-amzn — DISMISSED** [review]

**Review by chienhanlin — APPROVED** [review]

**Review by rajal-amzn — APPROVED** [review]

**Review by fierlion — APPROVED** [review]
